### ISSUE TITLE
Exception when using ScripArgs with >1 param in powershell

### ISSUE THREAD

**Issue by xavierpena** [cake-build/cake#1894]
### What You Are Seeing?

In the windows command line:

	>powershell -File test.ps1 -ScriptArgs '-param1="val1" -param2="val2"'
E:\svn-hydrique\DbManager\build\_tmp\test.ps1 : A parameter cannot be found that matches parameter name 'param2=val2''.
    + CategoryInfo          : InvalidArgument: (:) [test.ps1], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : NamedParameterNotFound,test.ps1

### What is Expected?

This part of the code should be reached in `test.cake` (it is not even built after the previous error):

	var invalidArgumentStr = "INVALID";
	var val1 = Argument("param1", invalidArgumentStr);
	var val2 = Argument("param2", invalidArgumentStr);

### What version of Cake are you using?

<package id="Cake" version="0.20.0" />

### Are you running on a 32 or 64 bit system?

64 bit system

### What environment are you running on?  Windows? Linux? Mac?

Windows

### Are you running on a CI Server?  If so, which one?

No

### How Did You Get This To Happen? (Steps to Reproduce)

Bat file:

	powershell -File test.ps1 -ScriptArgs '-param1="val1" -param2="val2"'
	PAUSE

Cake file:

	// ===
	// USAGE:
	// powershell -File test.ps1 -ScriptArgs '-param1="val1" -param2="val2"'

	// ===
	// CONFIGURATION:

	var invalidArgumentStr = "INVALID";
	var val1 = Argument("param1", invalidArgumentStr);
	var val2 = Argument("param2", invalidArgumentStr);

	// ===
	// SCRIPT:

	var target = Argument("target", "Default");

	Task("Publish")
		.Does(() =>
		{    
			// do nothing
		});

	Task("Default")
		.IsDependentOn("Publish");

	RunTarget(target);

**Comment by gep13** [comment]
Can you please confirm that you are using the latest Bootstrapper file?  I have just tried the .cake file that you provided (with a slight modification to output the input values) and I get the following...

When just calling build:

![image](https://user-images.githubusercontent.com/1271146/31988794-29ad1b3a-b968-11e7-9d93-d698fe000345.png)

i.e. I get INVALID

When calling with additional ScriptArgs:

![image](https://user-images.githubusercontent.com/1271146/31988816-438a94a6-b968-11e7-902a-a56dd34ad33f.png)


i.e. The arguments are passed through correctly.

**Comment by xavierpena** [comment]
I've fetched the bootstrapper again (via Visual Studio Code) like so:

Selected `Cake: install a bootstraper`, and choosing `PowerShell`.

This is the result when I try to run the .ps1 file in the VS console:

![image](https://user-images.githubusercontent.com/10937807/31989523-d28aa952-b972-11e7-99a0-16734b319de6.png)

**Comment by gep13** [comment]
In VSCode, looks like you are using the `cmd` terminal.  To get the command that you are trying to work to run, you are going to need to do it from a PowerShell terminal.

**Comment by xavierpena** [comment]
Please excuse my total lack of skills with powershell:

    param1: val1 -param2=val2
    param1: INVALID

![image](https://user-images.githubusercontent.com/10937807/31990238-ce43e9d8-b974-11e7-8f94-2098e4e07020.png)

**Comment by gep13** [comment]
Assuming that you have a PowerShell command window open (which I am guessing due to the blue background) you don't need to run `powershell` as it is already active.  You should be able to run this directly:

`.\test.ps1 -ScriptArgs '-param1-"val1" -param2="val2"'`

That aside though, I still can't reproduce what you are seeing 😢 

Okay, next questions...

What version of Cake are you using?
What OS are you on?
What version of PowerShell are you using?

**Comment by xavierpena** [comment]
> Assuming that you have a PowerShell command window open

Yes, you are correct. I did that and I don't know if the error I'm getting may shed some light on the issue:

![image](https://user-images.githubusercontent.com/10937807/31990949-d035c2dc-b976-11e7-99cf-8d9cc220f038.png)

> What version of Cake are you using?

<package id="Cake" version="0.23.0" />

> What OS are you on?

Windows 7 Professional

> What version of PowerShell are you using?

Major  Minor  Build  Revision
-----  -----  -----  --------
3      0      -1     -1

**Comment by gep13** [comment]
@xavierpena and what version of Cake are you using?

**Comment by xavierpena** [comment]
Sorry, I edited it right away. The "<" + ">" were hiding the text:

    package id="Cake" version="0.23.0"

**Comment by gep13** [comment]
@xavierpena and you get the same thing with 0.23.0 as well?

**Comment by xavierpena** [comment]
> you get the same thing with 0.23.0 as well?

Except for the original post (which used 0.20.0) all other comments show the behavior using 0.23.0.

> If you run &where.exe cake at your PowerShell prompt, what do you get back?

    INFO: Could not find files for the given pattern(s).

**Comment by gep13** [comment]
Ok, so that output means that you don't have Cake installed globally, so that is good.

Unfortunately, I am at a loss to suggest anything else...

Can you provide the exact command that you are running, and all the files that you are currently using, and I will try to replicate.

**Comment by xavierpena** [comment]
I hope the info provided below is enough, otherwise I can upload a .zip file somewhere with the exact files I am using:

	//// folder looks like this:

	* \tools (automatically downloaded by Cake)
	* test.ps1 (created with bootstrapper via Visual Studio Code)
	* test.cake (see code below)


	//// command line:

	powershell -File test.ps1 -ScriptArgs '-param1="val1" -param2="val2"'
	 
	(I specify "powershell" because I want to run it with a `test.bat` file, which works perfectly with a single argument)
	 

	//// test.cake:

	// ===
	// CONFIGURATION:

	var invalidArgumentStr = "INVALID";

	// For the "Build" task: input argument to specify if we are using a "Debug" build or a "Release" build;
	var configuration = Argument("buildconfiguration", "Debug");

	var val1 = Argument("param1", invalidArgumentStr);
	Console.WriteLine($"param1: {val1}");

	var val2 = Argument("param2", invalidArgumentStr);
	Console.WriteLine($"param1: {val2}");

	// ===
	// SCRIPT:

	var target = Argument("target", "Default");

	Task("Publish")
		.Does(() =>
		{    
			// do nothing
		});

	Task("Default")
		.IsDependentOn("Publish");

	RunTarget(target);

**Comment by gep13** [comment]
Can you also provide the bootstrapper that you have?  I want to make sure that it is the same also.

**Comment by xavierpena** [comment]
@gep13 Here it is (as I understand it, "boostrapper" == ".ps1 file")

https://pastebin.com/raw/Uiyr77qB

**Comment by mkiedrowski** [comment]
Shooul't parameters in powershell be separaed by comma?

**Comment by mkiedrowski** [comment]
Shouldn't parameters in PowerShell be separated by a comma? 
`.\test.ps1 -ScriptArgs '-param1-"val1"',  '-param2="val2"'`

**Comment by xavierpena** [comment]
@mkiedrowski That solves the issue indeed! Thanks a lot.

I was following [this](https://github.com/nunit/vs-project-loader/blob/master/build.cake#L39) `build.cake` example (and others, I think?) that stated that the usage was `'-param1-"val1" -param2="val2"'` instead of `'-param1-"val1"', '-param2="val2"'`.

**Comment by xavierpena** [comment]
@mkiedrowski 

Well, it worked out, but only partly:

	// Command line A:
	PS build-directory> .\ps1file.ps1 -ScriptArgs '-arg1="val1"', '-arg2="val2"'
	// Works OK

	// Command line B (batch file friendly):
	PS build-directory> powershell -File ps1file.ps1 -ScriptArgs '-arg1="val1"', '-arg2="val2"'
	// Fails: 
        // A parameter cannot be found that matches parameter name 'arg2=val2'.

My final goal is to have a .bat file with "Commant line B", so I'm still stuck.

**Comment by xavierpena** [comment]
I've finally found a way to solve the issue. It was quite convoluted, involving different factors.

Remember: my final goal is to use a bash (.bat) file. If you want to use directly the powershell console, all the following considerations might be useless for your case.

1. This is the correct way to pass arguments to powershell via bash file:

    powershell -File ps1file.ps1 -arg1 "val1" -arg2 "val2"

2. If you do this to call a Cake ps1, depending on the content of "val1" or "val2", you will have the following error:

> More than one build script specified

3. What are those cases in which "valN" is not valid? When it contains the point character "."

4. [The solution for that](https://github.com/cake-build/cake/issues/1248#issuecomment-249287930) is to use "--" instead of "-" before "argN":

    powershell -File ps1file.ps1 --arg1 "val1" --arg2 "val2"

**Comment by jusefb** [comment]
A little amendment to the above i had to add the equal sign:
`
powershell -File ps1file.ps1 --arg1="val1" --arg2="val2"
`
